### ISSUE TITLE
Decorator applied to base class instead of target class

### ISSUE THREAD

**Issue by LordZardeck** [loganfsmyth/babel-plugin-transform-decorators-legacy#60]
If i have a base class, extend it, then apply a decorator to the property of that new class, the base class and all the classes that extend from it have that decorator. It seems to be an issue from applying the decorator to the prototype, which is the base class. I'm not sure if this is a babel issue or this plugin's issue.

From the generated code:
```
_applyDecoratedDescriptor(_class2.prototype, 'logo', [_dec3], {
	enumerable: true,
	configurable: true,
	initializer: function initializer() {
		return '';
	}
})
```

`_class2` is the new class, and `_class2.prototype` is the base class. the decorator gets applied to the base class and all of it's children, adding the property to the children.

**Comment by loganfsmyth** [comment]
Would you be able to include an example of the original source code? It is hard to follow what you are describing.

**Comment by LordZardeck** [comment]
Without compiling it, this is the issue:

```
class Parent {
    
}

class Child1 extends Parent {
    @someDecorator('myvalue')
    prop = 'foobar';
}

class Child2 extends Parent {
    
}
```

`Child2` get's the decorator applied from `Child1` because the decorator is being applied to `Child1.prototype` which actually is `Parent`

If there is a quick way to create a fiddle like example i'll re-create it there. I can't share the code I have due to the company's IP policy

**Comment by loganfsmyth** [comment]
I'm happy to drop a snippet into my local setup if you make a snippet with some clear output. Not sure of any easy Fiddle that would handle decorators.

**Comment by LordZardeck** [comment]
I believe this replicates my issue:

```
function someDecorator( value ) {
	return function ( target, key ) {
		target.__storageProp = (target.__storageProp || {});

		target.__storageProp[ value ] =
			{
				key,
				value
			};
	}
}

function storedProps( target ) {
	target.prototype.showStoredProps = function () {
		console.log( Object.keys( target.prototype.__storageProp ) );
	}
}

@storedProps
class Parent {
	@someDecorator( 'base' )
	base = 'hello';
}

@storedProps
class Child1 extends Parent {
	@someDecorator( 'prop' )
	prop = 'foobar';
}

@storedProps
class Child2 extends Parent {
	@someDecorator( 'prop2' )
	prop2 = 'barfoo';
}
```

It's an extremely primitive version of the decorators, but I am able to replicate it at least. To see the effect:

```
var a = new Child2();
console.log(a.showStoredProps())
```

Will output: `["base", "prop", "prop2"]` and `prop` shouldn't be included

**Comment by loganfsmyth** [comment]
Your `target.__storageProp = (target.__storageProp || {});` is causing that. In the case where `target` is `Child1`, it will read `Child1.__storageProp` and it will already exist because `Child1.__proto__ === Parent`. If you want the storage to be per-constructor, you'll been to check `hasOwnProperty`.

    if (!target.hasOwnProperty('__storageProp')) target.__storageProp = {};

so it gets initialized for every constructor.

**Comment by LordZardeck** [comment]
'Doh I can't believe I didn't see that. I don't know what I was thinking. Sorry for the false Issue!

**Comment by loganfsmyth** [comment]
No problem :)
